inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <map>
+#include <unordered_set>
+
+#include <ie_metric_helpers.hpp>
+#include <ie_core.hpp>
+#include <threading/ie_executor_manager.hpp>
+#include <ie_algorithm.hpp>
+
+#include "auto_plugin.hpp"
+
+namespace AutoPlugin {
+namespace {
+    ConfigType mergeConfigs(ConfigType config, const ConfigType& local) {
+        for (auto && kvp : local) {
+            config[kvp.first] = kvp.second;
+        }
+        return config;
+    }
+
+    DeviceInformation SelectDevice(const std::vector<DeviceInformation>& metaDevices) {
+        for (auto& item : metaDevices) {
+            if (item.deviceName.find("CPU") == 0) {
+              return item;
+            }
+        }
+        IE_THROW(NotFound) << "No available device could be used";
+    }
+}  // namespace
+
+AutoInferencePlugin::AutoInferencePlugin() {
+    _pluginName = "AUTO";
+}
+
+IE::ExecutableNetworkInternal::Ptr AutoInferencePlugin::LoadExeNetworkImpl(const IE::CNNNetwork& network, const ConfigType& config) {
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+
+    // FIXME: always select CPU device now
+    DeviceInformation selectedDevice = SelectDevice(metaDevices);
+    IE::ExecutableNetwork executableNetwork;
+    try {
+        executableNetwork = GetCore()->LoadNetwork(network, selectedDevice.deviceName, selectedDevice.config);
+    } catch(const IE::Exception &iie) {
+        IE_THROW() << "Failed to load network to device named " << selectedDevice.deviceName
+                   << " with exception " << iie.what();
+    }
+
+    bool enablePerfCounters = false;
+    try {
+        enablePerfCounters =
+            executableNetwork.GetConfig(IE::PluginConfigParams::KEY_PERF_COUNT).as<std::string>() ==
+                IE::PluginConfigParams::YES;
+    } catch (...) {
+    }
+
+    return std::make_shared<AutoExecutableNetwork>(executableNetwork,
+                                                   selectedDevice,
+                                                   enablePerfCounters);
+}
+
+IE::QueryNetworkResult AutoInferencePlugin::QueryNetwork(const IE::CNNNetwork& network, const ConfigType& config) const {
+    IE::QueryNetworkResult queryResult = {};
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
```

Comment:
I suppose we should not pass config keys to AUTO?

Reply:
currently there is no configs for AUTO, but I think this logic is still necessary.